src/Microsoft.AspNetCore.WebHooks.Receivers/ApplicationModels/WebHookActionModelFilterProvider.cs

```diff
@@ -107,25 +113,39 @@ private void Apply(ActionModel action)
             }
 
             var filters = action.Filters;
+            AddReceiverFilters(action, filters);
             AddVerifyBodyFilter(action.Properties, filters);
             if (!_behaviorOptions.SuppressModelStateInvalidFilter)
             {
                 filters.Add(_modelStateInvalidFilter);
             }
         }
 
+        private void AddReceiverFilters(ActionModel action, IList<IFilterMetadata> filters)
+        {
+            if (action.Properties.TryGetValue(typeof(IWebHookFilterMetadata), out var filterMetadataObject) &&
+                filterMetadataObject is IWebHookFilterMetadata filterMetadata)
+            {
+                var context = new WebHookFilterMetadataContext(action);
+                filterMetadata.AddFilters(context);
+                foreach (var filter in context.Results)
+                {
+                    filters.Add(filter);
+                }
+            }
+        }
```

Comment:
Can this be simplified further? Who's adding to `action.Properties`? Can we just add the filter then?

Reply:
`WebHookActionModelPropertyProvider` has been adding to `ActionModel.Properties` forever. It does that mainly to provide inputs for the other 3 `IApplicationModelProvider` implementations, including this one. (I added a read of `ApplicationDescriptor.Properties` in `WebHookFilterProvider` but that's unique and should remain so.)

I'd rather not remove the use of `ActionModel.Properties` because it avoids duplicated code across these providers. But, as you suggest below, let's chat about it.

Reply:
In an offline conversation, we decided this approach was unusual but fine.